src/blog/articles/javascript-async-iterators/index.md

```diff
@@ -0,0 +1,689 @@
+
+
+### Converting a Node.js event emitter to an async iterable
+
+Another interesting use case for async iteration in Node.js is when dealing with repeated events happening over time.
+
+Events are generally fired by an _event emitter_ and, since version 12.16, Node.js offers an interesting utility to convert a sequence of events into an async iterable.
+
+We can see a simple example by using the third party module [`glob`](https://npm.im/glob) which allows us to find files matching a specific glob expression.
+
+In this example we will find and print all the JavaScript files (`.js` extension) in the current folder (and subfolders):
+
+```javascript
+import { on } from 'events'
+import glob from 'glob'
+
+const matcher = glob('**/*.js')
+
+for await (const [filePath] of on(matcher, 'match')) {
+  console.log(filePath)
+}
+```
+
+As you can see, we are using `on(matcher, 'match')` to create an async iterable that will _produce_ a new value every time the `matcher` instance fires a `match` event.
+
+Note that the value produced by this async iterable at every iteration is an array containing all the values contained in the original `match` event. This is the reason why we need to use destructuring to extract the `filePath`.
+
+> TODO: **add a section explaining that this loop is "endless" and mention the abort controller. Maybe provide an example as well**
+
+
+## Consuming paginated data with async iterators
+
+As we mentioned before with the DynamoDB examples, another great use case for async iteration is when we need to fetch data from a remote paginated dataset. Even more so when we cannot determine how to access the next page until we have fetched the previous one. This is a typical example of asynchronous sequential iteration and it's probably the most adequate use case for async iterators.
+
+Just to present a very simple example, let's use [a free and open-source Star Wars API](https://swapi.dev/) which allows us to access all the Star Wars characters in a paginated fashion.
+
+To get data from this API, we can make a GET request to the following endpoint:
+
+```text
+https://swapi.dev/api/people
+```
+
+This request will respond with a JSON message that looks like this:
+
+```json
+{
+	"count": 82,
+	"next": "http://swapi.dev/api/people/?page=2",
+	"results": [
+		{
+			"name": "Sly Moore",
+			"height": "178",
+      "...": "mode fields...",
+    },
+    {
+      "name": "Another character",
+			"height": "whatever",
+      "...": "mode fields...",
+    },
+    {
+      "...": "more characters"
+    }
+  ]
+}
+```
+
+Note that the field `next` contains the URL that we can use to fetch the data from the following page. All the records for the current page are presented in the `results` field.
+
+With these details in mind, this is how we can create a custom client that allows us to fetch all the characters using the `for await...of` syntax:
+
+```javascript
+import axios from 'axios'
+
+function starWarsCharacters () {
```

Comment:
Generators are already iterable. No need to create the iterable function explicitly.